File: exercism/java/rotational-cipher/src/main/java/RotationalCipher.java
import java.util.HashMap;
import java.util.Map;

class RotationalCipher {

    static final String alphabet = "abcdefghijklmnopqrstuvwxyz";
    final String cipherAlphabet;
    private final Map<Character, Character> alphabetToCipher = new HashMap<>();

    RotationalCipher(int shiftKey) {
        shiftKey = shiftKey % 26;
        this.cipherAlphabet = alphabet.substring(shiftKey) + alphabet.substring(0, shiftKey);
        for (int i = 0; i < alphabet.length(); i++) {
            alphabetToCipher.put(alphabet.charAt(i), cipherAlphabet.charAt(i));
        }
    }

    String rotate(String data) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < data.length(); i++) {
            if(Character.isUpperCase(data.charAt(i))){
                result.append(Character.toUpperCase(
                        alphabetToCipher.get(Character.toLowerCase(data.charAt(i)))));
            }
            else{
                result.append(
                        alphabetToCipher.getOrDefault(Character.toLowerCase(data.charAt(i)),data.charAt(i)));
            }
        }
        return result.toString();
    }

}

File: exercism/clojure/rotational-cipher/src/rotational_cipher.clj
(ns rotational-cipher)


(defn ->cipher-alphabet
  "Rotate a string XS-string by k toward the left
  Eg: (rotate 'abcd' 2) -> 'cdab'
  "
  [xs-string, k]
  (let [n (count xs-string)
        k (mod k n)]
    (str (subs xs-string k) (subs xs-string 0 k))))

(defn rotate [s k]
  (let [cipher (zipmap "abcdefghijklmnopqrstuvwxyz" (->cipher-alphabet "abcdefghijklmnopqrstuvwxyz" k))]
    (->> s
         (map (fn [c-char]  (if (Character/isLowerCase c-char) (get cipher (Character/toLowerCase c-char) c-char)
                                (Character/toUpperCase (get cipher (Character/toLowerCase c-char) c-char)))))
         (clojure.string/join))))